src/NHibernate/Type/TimestampType.cs

```diff
@@ -1,142 +1,50 @@
 {
 	/// <summary>
-	/// This is almost the exact same type as the DateTime except it can be used
-	/// in the version column, stores it to the accuracy the database supports, 
-	/// and will default to the value of DateTime.Now if the value is null.
+	/// This is almost the exact same type as the <see cref="DateTimeType" /> excepted it can be used
+	/// in the version column, stores it to the accuracy the database supports
+	/// and will default to the value of <see cref="TimestampType.Now" /> if the value is not valid.
 	/// </summary>
 	/// <remarks>
-	/// <p>
+	/// <para>
 	/// The value stored in the database depends on what your data provider is capable
-	/// of storing.  So there is a possibility that the DateTime you save will not be
-	/// the same DateTime you get back when you check DateTime.Equals(DateTime) because
+	/// of storing. So there is a possibility that the DateTime you save will not be
+	/// the same DateTime you get back when you check <see cref="DateTime.Equals(DateTime)" /> because
 	/// they will have their milliseconds off.
-	/// </p>  
-	/// <p>
-	/// For example - SQL Server 2000 is only accurate to 3.33 milliseconds.  So if 
+	/// </para>
+	/// <para>
+	/// For example - SQL Server 2000 is only accurate to 3.33 milliseconds. So if
 	/// NHibernate writes a value of <c>01/01/98 23:59:59.995</c> to the Prepared Command, MsSql
 	/// will store it as <c>1998-01-01 23:59:59.997</c>.
-	/// </p>
-	/// <p>
+	/// </para>
+	/// <para>
 	/// Please review the documentation of your Database server.
-	/// </p>
+	/// </para>
+	/// <para>
+	/// If you are looking for the most accurate date and time storage accross databases use the
+	/// <see cref="TicksType"/>.
+	/// </para>
 	/// </remarks>
 	[Serializable]
-	public partial class TimestampType : PrimitiveType, IVersionType, ILiteralType
+	public class TimestampType : AbstractDateTimeType
 	{
-		/// <summary>
-		/// Retrieve the current system Local time.
-		/// </summary>
-		/// <value>DateTime.Now</value>
-		protected virtual DateTime Now => DateTime.Now;
-
-		/// <summary>
-		/// Returns the DateTimeKind for type <see cref="TimestampUtcType"/>.
-		/// </summary>
-		/// <value>Returns DateTimeKind.Unspecified</value>
-		protected virtual DateTimeKind Kind => DateTimeKind.Unspecified;
-
-		public TimestampType() : base(SqlTypeFactory.DateTime)
-		{
-		}
+		/// <inheritdoc />
+		public override string Name => "Timestamp";
 
-		public override object Get(DbDataReader rs, int index, ISessionImplementor session)
-		{
-			return DateTime.SpecifyKind(Convert.ToDateTime(rs[index]), Kind);
-		}
-
-		public override object Get(DbDataReader rs, string name, ISessionImplementor session)
-		{
-			return Get(rs, rs.GetOrdinal(name), session);
-		}
-
-		public override System.Type ReturnedClass
-		{
-			get { return typeof(DateTime); }
-		}
+		/// <inheritdoc />
+		protected override DateTime GetDateTimeToSet(object value, ISessionImplementor session) =>
+			(value as DateTime?) ?? Now;
```

Comment:
This override should be removed if we want to clean the failed attempt at replacing null values by `Now`. By the way if keeping it, we should probably call `Seed` instead of `Now`.